retrieval/manager.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retrieval
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Appendable returns an Appender.
+type Appendable interface {
+	Appender() (storage.Appender, error)
+}
+
+// NewScrapeManager is the ScrapeManager constructor
+func NewScrapeManager(logger log.Logger, app Appendable) *ScrapeManager {
+
+	return &ScrapeManager{
+		append:        app,
+		logger:        logger,
+		actionCh:      make(chan func()),
+		scrapeConfigs: make(map[string]*config.ScrapeConfig),
+		scrapePools:   make(map[string]*scrapePool),
+		graceShut:     make(chan struct{}),
+	}
+}
+
+// ScrapeManager maintains a set of scrape pools and manages start/stop cicles
+// when receiving new target groups form the discovery manager.
+type ScrapeManager struct {
```

Comment:
Not sure if this makes sense, but it feels to me like the discovery package should provide an interface that other packages can implement to handle recv'ing  of target groups and then implementation details of how the consumer wants to use those groups is all the other package needs to do. Maybe that's actually what we're doing and I'm misreading?

Reply:
can you give an example of maybe a simple gist with your idea.

Reply:
sure, I'll give it a shot, or perhaps we can discuss on irc?

Reply:
sure , fork it  and change it to your idea and can discuss.